kernel/dev/pci.c

```diff
@@ -0,0 +1,344 @@
+#include <dev/pci.h>
+#include <acpi/acpi.h>
+#include <mm/vmm.h>
+#include <lib/vector.h>
+#include <lib/print.h>
+#include <sys/cpu.h>
+
+struct mcfg_entry {
+    uint64_t mmio_base;
+    uint16_t segment;
+    uint8_t  start;
+    uint8_t  end;
+    uint32_t reserved;
+};
+
+union __attribute__((packed)) msi_address {
+    struct {
+        uint32_t reserved0    : 2;
+        uint32_t dest_mode    : 1;
+        uint32_t redir_hint   : 1;
+        uint32_t reserved1    : 8;
+        uint32_t dest_id      : 8;
+        uint32_t base_address : 12;
+    };
+
+    uint32_t raw;
+};
+
+union __attribute__((packed)) msi_data {
+    struct {
+        uint32_t vector    : 8;
+        uint32_t delivery  : 3;
+        uint32_t reserved  : 3;
+        uint32_t level     : 1;
+        uint32_t trig_mode : 1;
+        uint32_t reserved0 : 16;
+    };
+
+    uint32_t raw;
+};
+
+uint32_t (*pci_read)(struct pci_device*, uint32_t, int);
+void (*pci_write)(struct pci_device*, uint32_t, uint32_t, int);
+
+static VECTOR_TYPE(struct mcfg_entry) mcfg_entries = VECTOR_INIT;
+static VECTOR_TYPE(struct pci_device*) devlist = VECTOR_INIT;
+static void scan_bus(uint8_t bus);
+
+static uint32_t mcfg_read(struct pci_device* dev, uint32_t offset, int access_size) {
+    VECTOR_FOR_EACH(&mcfg_entries, ent,
+	if (dev->seg != ent->segment)
+            continue;
+	else if ((dev->bus > ent->end) || (dev->bus < ent->start))
+            continue;
+
+	uintptr_t target = ((dev->bus - ent->start) << 20) | (dev->slot << 15) | (dev->func << 12);
+	target += ent->mmio_base + offset + VMM_HIGHER_HALF;
+
+        switch (access_size) {
+            case 1:
+                return *(uint8_t*)target;
+            case 2:
+                return *(uint16_t*)target;
+            case 4:
+                return *(uint32_t*)target;
+	}
+    );
+
+    return 0;
+}
```

Comment:
right side pointers, additionally please re-indent the file with 4 tabs and use braces around if statements